grr/server/grr_response_server/gui/ui/components/flow_details/plugins/osquery_details.ts

```diff
@@ -0,0 +1,71 @@
+import { Component, ChangeDetectionStrategy, OnInit, OnDestroy } from '@angular/core';
+import {Plugin} from './plugin';
+import { map, tap, flatMap, takeUntil, filter } from 'rxjs/operators';
+import { FlowState, FlowResultSet, FlowResult, FlowListEntry } from '@app/lib/models/flow';
+import { Observable, Subject, from } from 'rxjs';
+import { OsqueryResult, OsqueryArgs } from '@app/lib/api/api_interfaces';
+
+/**
+ * Component that displays the details (status, errors, results) for a particular Osquery Flow.
+ */
+@Component({
+  selector: 'osquery-details',
+  templateUrl: './osquery_details.ng.html',
+  styleUrls: ['./osquery_details.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class OsqueryDetails extends Plugin implements OnDestroy {
+  private readonly onDestroyHappened$ = new Subject<void>();
+
+  hasLoadedResults = false;
+
+  flowRunning$ = this.flagByState(FlowState.RUNNING);
+  flowCompleted$ = this.flagByState(FlowState.FINISHED);
+  flowError$ = this.flagByState(FlowState.ERROR);
+
+  args$: Observable<OsqueryArgs> = this.flowListEntry$.pipe(
+      map((flowListEntry) => flowListEntry.flow.args as OsqueryArgs)
+  );
+
+  osqueryResult$: Observable<OsqueryResult> = this.flowListEntry$.pipe(
+    flatMap((listEntry) => listEntry.resultSets),
+    flatMap((singleResultSet) => singleResultSet?.items),
+    map((singleItem) => singleItem?.payload as OsqueryResult)
```

Comment:
This casts nullable values to a non-nullable type, which can trigger hard-to-trace runtime errors. Instead of casting away the `undefined` and `null`, please use `filter(isNonNullable)`.

Alternatively, cast to `OsqueryResult | undefined` and use this type on the member definition as well.

Reply:
You should be able to remove `?` from `singleItem?.payload` after `filter(isNonNull)`.